ssl/record/rec_layer_s3.c

```diff
@@ -1554,30 +1554,44 @@ int ssl3_read_bytes(SSL *s, int type, int *recvd_type, unsigned char *buf,
         return -1;
     }
 
-    /*
-     * If we've sent a close_notify but not yet received one back then ditch
-     * anything we read.
-     */
     if ((s->shutdown & SSL_SENT_SHUTDOWN) != 0) {
-        /*
-         * In TLSv1.3 this could get problematic if we receive a KeyUpdate
-         * message after we sent a close_notify because we're about to ditch it,
-         * so we won't be able to read a close_notify sent afterwards! We don't
-         * support that.
-         */
-        SSL3_RECORD_set_length(rr, 0);
-        SSL3_RECORD_set_read(rr);
-
         if (SSL3_RECORD_get_type(rr) == SSL3_RT_HANDSHAKE) {
             BIO *rbio;
+            /*
+             * We need to check the message type of the received data. We're not
+             * "in init" otherwise we won't have got here - so this is a new
+             * handshake message. However it's theoretically possible for us to
+             * have received part of the message header before us sending
+             * close_notify, and for us to be receiving the remainder of it now.
+             */
+            unsigned char mt = s->rlayer.handshake_fragment_len > 0
+                               ? s->rlayer.handshake_fragment[0]
+                               : *SSL3_RECORD_get_data(rr);
```

Comment:
I think handshake_fragment_len == 0 if type == SSL3_RT_HANDSHAKE, according to a comment above in the function.

Reply:
A later comment then says: s->rlayer.handshake_fragment_len == 4  iff  rr->type == SSL3_RT_HANDSHAKE

Clearly I don't understand this code good enough.

Reply:
```type``` is the type of data that the caller is requesting; in the cases considered here it will be application data.
(The actual record we're processing would be a handshake message, though, of course.)

I'm not 100% conversant with this code, either, but I do not really believe that the comment you quote is accurate at the location it is currently placed, since we populate handshake_fragment later in the function at present.

Reply:
As @kaduk says the comment you refer to only applies if `type == SSL3_RT_HANDSHAKE` which is only true if the caller is asking for handshake data - which isn't the case here.

> I'm not 100% conversant with this code, either, but I do not really believe that the comment you quote is accurate at the location it is currently placed, since we populate handshake_fragment later in the function at present.

The function loops. If we receive handshake data that is less than `SSL3_HM_HEADER_LENGTH` in length then we go back to `start` (see the comment "fragment was too small"). It then attempts to get another record. That could of course end up returning to the calling application if no record is available. At that point the application could send `close_notify`. If another read then occurs and more handshake data has arrived then we will get to this point in the code and `handshake_fragment_len` will be > 0.